fix #152 improve documentation

vinnymac opened the pull request:
The goal here is to improve documentation. Both the documentation and the user experience around that documentation require changes.

* [x] Use gitbook to keep documentation style consistent across redux projects ([redux](redux.js.org) and [redux-saga](https://redux-saga.js.org))
* [x] Add TOCs
* [x] Make the docs less long and unruly
* [x] Close #172 documentation issue
* [x] Close #204 documentation issue
* [x] Add a Beginner Tutorial with Demo and Installation instructions
* [x] Use [js.org](https://github.com/js-org/dns.js.org) so that people can access the docs via `redux-actions.js.org`

[New Documentation Demo](https://vinnymac.gitbooks.io/redux-actions/content/)

yangmillstheory commented:
Are those checkboxes still outstanding? Let me know when I should start review (might be not be until this weekend on my end).

yangmillstheory commented:
Also is there a URL where I can see the docs site?

vinnymac commented:
@yangmillstheory yea this is just a starting point for the PR. Nowhere near finished. Here is where you can find the [new docs site](https://vinnymac.gitbooks.io/redux-actions/content/).

vinnymac commented:
@yangmillstheory if you want to review it you can start reviewing the new documentation site. I would avoid the `Beginner Tutorial` for now, since I may not have the time to finish that part just yet. I added a lot of new examples to the API Reference as well as tweaking the existing ones. Let me know if you think we can make any improvements.

I also added an example for #204 [here](https://vinnymac.gitbooks.io/redux-actions/content/docs/api/#handleactiontype-reducer-defaultstate).

I also added an example for `metaCreator` which was missing.

Let me know your thoughts on the format I chose. I think it might be nice if we add typings to all the parameters below each method header.

yangmillstheory commented:
This looks great; love the beginner tutorial as well (can we just call it "Tutorial"?)  🎉 

I had a couple of comments:

* [ ] I think the page title should be "Redux Actions", right now it's "GitBook"
* [ ] [the CHANGELOG is empty](https://vinnymac.gitbooks.io/redux-actions/content/docs/ChangeLog.html)
* [ ] Can we have subsections for the API reference? Maybe group them by actions vs. reducers, or have a single subsection for each API?
* [ ] not required, but what do you think of having a section up front where we list all the contributors?

This is a great starting point for the documentation site, and 500X better than what we currently have.

vinnymac commented:
@yangmillstheory for me the page title says redux-actions, not sure why it says `GitBook` for you, will have to look into it. 

The `ChangeLog` looks like this for me, where it is pointing to the releases page for changes.
<img width="953" alt="screen shot 2017-07-01 at 3 51 47 pm" src="https://user-images.githubusercontent.com/1832781/27765004-39bbc100-5e75-11e7-9a0e-b2e66e603f92.png">

When you say `subsection` do you mean a specific pages dedicated to a subject? For example
2. API Reference
  2.1 Actions
  2.2 Reducers

We can definitely make additions like that if it will help readers.

I can add a dedicated contibutors page as number 6 if that is what you mean. Would it just be a list of names, or do we want to do something fancy with photos?

I can have all this done this weekend so we can start iterating. Not sure who to ask about getting `http://redux-actions.github.io` setup, but I think we need to do that [first](https://github.com/js-org/dns.js.org#step-2) before we can get our `redux-actions.js.org` PR in.

yangmillstheory commented:
Let's do option 2 for the API reference, that way we can expand each section with more examples if needed, without it becoming too big. Also nothing fancy with the list of contributors.


***

Based on what `redux-saga` is doing, they have all their docs in the `gh-pages` branch; this sets up the `.github.io` DNS. 

For our case, I think we can do the same thing on a `gh-pages` branch (I can set that up and you can merge this into that branch; `master` README would also have to be cleaned up).

But I actually don't have enough access to this repository to set that up; I'll e-mail @acdlite regarding this.

For now, can we get unblocked by using `https://vinnymac.gitbooks.io/redux-actions/` in the PR to js.org, as in https://github.com/js-org/dns.js.org/blob/master/cnames_active.js#L670?

vinnymac commented:
@yangmillstheory https://redux-actions.js.org/

yangmillstheory commented:
What do you mean that the DNS still needs to be configured? https://github.com/js-org/dns.js.org/pull/1333 looks like it created the CNAME to your gitbook URL.

Later we can self-host the site using GitHub Pages (I emailed @acdlite and CC'ed you). For now, thanks for hosting. 

I can take another look tonight; is there anything else you would like to include?

vinnymac commented:
Gitbook was giving me an error when I configured it. But maybe you are right. In my experience you usually have to wait a little while for the records to update, so I figure that is why gitbook is upset.

Doing one last look over can't hurt. Thanks for all the help by the way. It has been greatly appreciated.

yangmillstheory commented:
No problem; this is a huge win for the project, thanks for taking the initiative.

yangmillstheory commented:
Does the Edit functionality work now?

yangmillstheory commented on build/gitbook.css:
```diff
@@ -0,0 +1,3 @@
+.token.operator {
+  background: none;
+}
```
Should this folder be checked in?

yangmillstheory commented on package.json:
```diff
@@ -17,7 +17,12 @@
+    "docs:prepare": "gitbook install",
+    "docs:build": "npm run docs:prepare && gitbook build -g acdlite/redux-actions",
+    "docs:watch": "npm run docs:prepare && gitbook serve",
+    "docs:publish": "npm run docs:clean && npm run docs:build && cp CNAME _book && cd _book && git init && git commit --allow-empty -m 'update book' && git checkout -b gh-pages && touch .nojekyll && git add . && git commit -am 'update book' && git push git@github.com:acdlite/redux-actions gh-pages --force"
```
Does this need to be run separately from `npm publish`? Can we use the [`postpublish` hook](https://docs.npmjs.com/misc/scripts) to run this?

vinnymac replied:
Well I guess technically just because you are releasing a new version of the library on npm doesn't necessarily mean you are releasing new documentation. If you want it to be part of that process though I don't see why it couldn't be added to `postpublish`. I added that more for convenience in case you wanted it on `gh-pages` like we planned.

yangmillstheory replied:
OK it's up to you, we can leave it out for now. By the way, how will this work if I want to publish new documentation, since it's hosted in your gitbook URL? Do we need to grant access to developers of this repo until @acdlite sets up pushing to GitHub pages (I don't have access to the repository settings)?

vinnymac replied:
I have given both you and @acdlite full administrative privileges over the gitbook. We will likely want to get it done the other way asap so others can be able to contribute without requiring access.

yangmillstheory replied:
I agree; it's up to him to respond to the email at this point.

yangmillstheory commented:
@vinnymac just a couple of questions:

* https://github.com/acdlite/redux-actions/pull/222/files#r125202994
* https://github.com/acdlite/redux-actions/pull/222#discussion_r125203205

vinnymac commented:
The Edit functionality is pointing to `master` right now, so that won't work unless we merged it to master. I could have it point to my fork, but I don't think that is what we want. So we either merge it to master or we merge it to a separate branch and link to that one.

yangmillstheory commented:
OK, will merge by midnight.
On Sun, Jul 2, 2017 at 8:51 PM Vincent Taverna <notifications@github.com>
wrote:

> The Edit functionality is pointing to master right now, so that won't
> work unless we merged it to master. I could have it point to my fork, but I
> don't think that is what we want. So we either merge it to master or we
> merge it to a separate branch and link to that one.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/acdlite/redux-actions/pull/222#issuecomment-312543985>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ACmkd39wil6S6NUrlwoucuzVUG7JyKIUks5sKGVUgaJpZM4OIC86>
> .
>
-- 
Victor Alvarez

yangmillstheory commented:
Hm, some of the links are broken - try clicking around on the documentation site from the home page.

yangmillstheory commented:
@vinnymac I just pushed the fixes to `master`; how can I push the changes to your gitbook?

vinnymac commented:
@yangmillstheory somehow the links were pointing to `/api` instead of `/docs/api`. Not sure what happened exactly, but I ended up updating `vinnymac/redux-actions/tree/master` to the latest and pointing gitbook at that branch and it seems to have fixed the issue for me.

I could always transfer ownership of the entire gitbook to you. But I can't seem to find a way to switch from one repository to another because I don't see `acdlite/redux-actions` in the list on the settings page. If you navigate [here](https://www.gitbook.com/book/vinnymac/redux-actions/settings/github) can you try to link it to this repository instead of mine? May be a permissions thing.

yangmillstheory commented:
Thanks. I can't connect your GitBook to this repo (don't have access to the settings page, and the integration is configured there).

I guess we'll just have to wait for a response from @acdlite. It's the holiday after all though. So it looks like the current fix is to fix it your repo, and merge the changes back here (won't do anything but keep things in sync).

yangmillstheory commented:
@vinnymac for now, is there a place others can submit documentation PRs to to update the redux-actions.js.org?

timche commented:
@yangmillstheory try to contact him on [Twitter](https://twitter.com/acdlite). I assume he has muted this repo.

vinnymac commented:
@yangmillstheory if you make a PR to the master branch of vinnymac/redux-actions it will automatically update the docs once merged. Until acdlite responds I guess that is just how it will have to be done.

yangmillstheory commented:
Will do tim, I actually emailed him at the address on his profile.
